### PR TITLE
add on batch method

### PR DESCRIPTION
**Pull request by nicofirst1**
Refer to [this issue](https://github.com/facebookresearch/EGG/issues/175)

### CODE REVIEW

**Comment by nicofirst1** [comment]
[This](https://github.com/facebookresearch/EGG/pull/176#issuecomment-740863709)

**Comment by nicofirst1** [comment]
Actually, after some search, I found out that the [issue](https://github.com/NVIDIA/sentiment-discovery/issues/63) is in the torch library. I added another catch here 58c3ed2

File: egg/core/callbacks.py [hunk 1]
```diff
@@ -44,6 +44,9 @@ def on_epoch_begin(self, epoch: int):
     def on_epoch_end(self, loss: float, logs: Interaction, epoch: int):
         pass
 
+    def on_batch(self, logs: Interaction, loss: float, data_len: int):
```
Comment (robertodessi) [inline egg/core/callbacks.py:47]: what's data_len here, is it needed? If it's for the progress bar use case I would have this as a field of the hypothetical ProgressBarUpdater callback rather than parameter here. Especially because it's the same for every call to on_batch.
I would add a batch_id instead.
And how about changing the name to on_batch_end?

File: egg/core/trainers.py [hunk 2]
```diff
@@ -189,6 +189,12 @@ def train_epoch(self):
 
         self.optimizer.zero_grad()
 
+        try:
+            data_len = len(self.train_data)
+        except (TypeError, AttributeError):
+            # len not implemented
+            data_len = -1
+
```
Comment (robertodessi) [inline egg/core/trainers.py:197]: se previous comment, I don't think we need this here. Games having a callback could do that on their own

File: egg/core/trainers.py [hunk 3]
```diff
@@ -231,6 +237,8 @@ def train_epoch(self):
                 interaction = Interaction.gather_distributed_interactions(interaction)
             interaction = interaction.to("cpu")
             interactions.append(interaction)
+            for callback in self.callbacks:
+                callback.on_batch(interaction, optimized_loss.detach(), data_len)
```
Comment (robertodessi) [inline egg/core/trainers.py:241]: don't think detach() is needed here, is it?

Comment (nicofirst1) [reply egg/core/trainers.py:241]: It depends on the user I guess, If you want to perform some operations on the loss in numpy you need it to be detached. But you can probably do it in the callback

**Comment by robertodessi** [comment]
Apart from the comments, it looks good, thanks! :)

File: egg/core/callbacks.py (hunk 1 above)
Comment (nicofirst1) [reply egg/core/callbacks.py:47]: > ProgressBarUpdater callback rather than parameter here
You're right

>I would add a batch_id instead
What would be the need for it?

> And how about changing the name to on_batch_end?
Approved

Comment (robertodessi) [reply egg/core/callbacks.py:47]: Printing stats for the current batch or, for instance, using it to update a progressbar callback.
Same use cases of epoch_id

File: egg/core/trainers.py (hunk 3 above)
Comment (robertodessi) [reply egg/core/trainers.py:241]: yes I would leave it to the user since we're not saving much memory anyway. and we shouldn't assume any training operation is done on the loss in the callback since it won't be backprop through anyway
